### ISSUE TITLE
Cheating on large scale

### ISSUE THREAD

**Issue by g964** [Codewars/codewars.com#214]
I found that somebody (I can't call him a Codewarrior though he is 2kyu) passed **all** his katas in Ruby by cheating. 
The many different ways he is using have all the same aim: `==` will always return true. I could give some "examples" of his cheating but I don't want to give them publicly.

**Comment by Drymonade** [comment]
I saw him too. Is there any way to delete his "solutions"?

**Comment by g964** [comment]
They can't be deleted by me, only painfully invalidated:-(

**Comment by jhoffner** [comment]
Which user is this?

Sent from Outlook




On Fri, Dec 11, 2015 at 2:24 AM -0800, "g964" <notifications@github.com> wrote:












They can't be deleted by me, only painfully invalidated:-(

**Comment by Drymonade** [comment]
http://www.codewars.com/users/danielpclark

**Comment by bkaestner** [comment]
@jhoffner, while we're at it, how should I provide you [some information about this loophole](https://gitter.im/bkaestner/codewars-rules?at=565f60c62488cc807874b8fa)?

**Comment by g964** [comment]
Seems that there are at least two "loophole"...

**Comment by bkaestner** [comment]
Well, @danielpclark also gave [some information on the "Cheat Detection" page](http://www.codewars.com/topics/528184e02a901d9d9400006a#56682740003a22bfc0000063). Funny enough, he mentions some things I've been preaching for a while, e.g.

> - Tests being ordered so a person can just set a global variable counter and write the code responses based on the error messages.

That's something I very often "cheated" on in order to show flaws in Haskell beta katas (yes, there are [global counters in Haskell](http://www.codewars.com/kata/reviews/5467d5462edf8bbf40000141/groups/5469c7601f3bb42b490003cd)) and introduce random tests to the environment. See also solutions like [this](http://www.codewars.com/kata/reviews/551915b023d4adc218000093/groups/554287575b89d1c347000048) or [ŧhis](http://www.codewars.com/kata/reviews/542d0694d5847818080000f1/groups/542e7247734f7ddf5f000628).

> There is no such thing as cheating when it comes to TDD. You just write more stringent specs as you go along.

From a TDD point, yeah. You write a test, you write code that passes the test, you write the next test. That's what TDD is about, after all, regardless of the language. What's important is to use the tools in the box to have _meaningful tests_. Do not only use `assert_equals`, but also `assert_not_equals` (or your language equivalent) in order to have some tests that use a negative predicate (should probably add this to the [test section](https://bkaestner.github.io/codewars-rules/#tested-testing-theories)).

However, changing the semantics of the testing framework is like taking an exam, erase all the assignments and give yourself a grade (A+, 1, 10, or whatever is "perfect" in your school system). Sure, all the tests pass, but they're not meaningful anymore.

Either way, since he has a GitHub account, he's probably going to get a notification by this post.

**Comment by danielpclark** [comment]
I wouldn't mind having my points revoked.  I solved everything properly for the first 315 points.  Then I found it an interesting challenge to craft a better test passer.  Yes as a game that would be considered a cheat so I agree with you there.  Ideally I want to help people write better test scenarios and raise awareness of Ruby and how to code better.  The last 1000 points were me working on variations of the so called "cheat".  So admittedly I'm fine having points revoked but I would like it if the answers were left up for people to implement their tests until it goes away.

> I found that somebody (I can't call him a Codewarrior though he is 2kyu) passed all his katas in Ruby by cheating. 

@g964 It wasn't **all** but it was many.  Just around 1000 points worth.

I have found about 5% of the challenges I did had no way for me to guess pass its testing design.  I applaud them for their test design.

For me I'm more interested in the challenges and learning as much about the programming language(s) as I can.  These last 1000 points were about testing a challenge I saw about getting a viable copy-paste solution.  I named them various things but the gist of it is don't use only assert_equals.

Again I am passionate about programming and I'm all for having those last 1000 points removed.  If the DB manager needs a time frame they were entered to remove them they were between Dec 9th to 11th.  But like I said I want to encourage a greater understanding of the language and encourage better testing practices.

I have 3 code kata challenges of my own posted there.  Feel free to get them to pass by any means necessary ;-)

**Comment by g964** [comment]
I like your frankness and I must say that I was taught (laboriously:-) a bit more of Ruby in order to invalidate your "solutions"... But I would like to know what you would have done if nobody has discovered your "practice". Maybe Codewars could hire you so that you can fight cheating:-)

**Comment by danielpclark** [comment]
@g964 Haha yes! :-) That sounds like a wonderful grey-hat to white-hat kind of story.

As far as what I would have done, or will do, I feel I've exhausted the route of solution I was exploring.  It doesn't make me feel as clever anymore and I'm more interested in a challenge.  The old path is no longer a challenge for me.

I love teaching and educating others.  It would be nice to encourage the CodeWars system/staff to have some default hacks tested against all new entries and caution the user about not testing thoroughly enough.  Maybe I could contribute my efforts to that ;-)

**Comment by g964** [comment]
Is somebody acting in this case?

**Comment by raulbc777** [comment]
I found a cheating solution in a kata I authored:

"Find the First Number in Having a Certain Number of Divisors I (Ruby)"
http://www.codewars.com/kata/5612ab201830eb000f0000c0/solutions/ruby

It was reported publicly by the user: @mbr84.

The author of this cheating solution is the user @danielpclark .

I read above what @danielpclark  wrote and I understand his point. No doubt that his knowledege about hacking would have been very valuable. 

From my point of view with "his love to teach and educate others", making his cheating solutions publicly is creating a disturbing issue in this community. 

If he would have created these cheating solutions and sent them to @jhoffner and to other leaders of CW just to warn and improve the cheating detecting systems, we would not have doubt the respect that this user should deserve.

What are we having now?

1. Many people saying: "What's going on with some solutions?
2. We are having free classes given by danielpclark how to crack the cheating detection system.

I think that many times we may do or make many things but we should ask ourselves before performing them: Is this correct or not?

A border line was traspassed here.

**Comment by bkaestner** [comment]
Note that there is no disclosure specific section in [the terms](http://www.codewars.com/about/terms-of-service) or [the about section](http://www.codewars.com/about). So @danielpclark didn't really have a platform to tell @Codewars cheats. Maybe there are some plans to include them in #187, but I have no idea.

That being said, I have some bad news. I've found two cheats. One of them is trivial to run, but obvious (and _might_ get caught by the cheat detection), the other is a little bit of a mess, but provides an undetectable way to solve a kata without actually doing any work.

Still waiting for some information about the disclosure rules. But I guess I'll write a short article about the general problem that lies behind both of the "bugs".

**Comment by danielpclark** [comment]
One great way authors may prevent the assert_equals kind of exploit is to un-define both `==` and `method_missing` and re-implement `super`.  So for every object being tested we could do.

```ruby
def fix_things_to_test( thing )
  thing.tap {
    undef :example, :example_missing
    def example other
      super
    end

    def example_missing *a, &b
      super
    end
  }
end
```

This could be safe for the kata suite in every situation except for those that are designed for the user to mod those methods in the first place.

I propose adding a `Test#protect_methods` internal feature.  The purpose would be to preserve the languages original method behavior for specific methods.  So the authors would set it up with this at the begging of their test suite.

```ruby
Test.protect_methods :==, :method_missing, :to_f, :etc
Test.undef_methods :to_i, :to_s
```

This is just an example scenario where everything being tested would have the original methods restored via the `protect_methods` and the other methods undefined automatically in the test with the `undef_methods` feature.

Then the test suite can inquire these two internal variables before each object gets evaluated and execute the meta-programming method re-assignment.  It would be simple to loop over each method in these variables and undef/def them to `super`.

Thoughts?

**Comment by bkaestner** [comment]
> PS: Ruby should always un-define these few items (back-tick, system, exec)
>
> ```ruby
> undef :`, :system, :exec
> ```

Some katas depend on that functionality. Basically the same problem as for `==`:

> This could be safe for the kata suite in every situation except for those katas that are designed for the user to [mod] those in the first place.

**Comment by g964** [comment]
Interesting ideas but two remarks:

1- a bit too vague. You should publish these ideas on Github as a kind of guide with full examples.
It would offset the difficulties that you are creating to the community and morever it would better your CV:-)

2- Protecting Ruby against cheaters is becoming a heavy task for authors. 
I doubt very much that, from now on, somebody will write Ruby katas! 
If @jhoffner doesn't take a serious action - ie your disqualification - cheating will spread like a plague
all the more that everybody will now have a lot of ways to cheat. 
It will reduce to zero all the brilliant effort of the CW staff who made Codewars unique.

**Comment by g964** [comment]
Interesting ideas but two remarks:

1- a bit too vague. You should publish these ideas on Github as a kind of guide with full examples.
It would offset the difficulties that you are creating to the community and morever it would better your CV:-)

2- Protecting Ruby against cheaters is becoming a heavy task for authors. 
I doubt very much that, from now on, somebody will write Ruby katas! 
If @jhoffner doesn't take a serious action - ie your disqualification with the removal of your solutions or the suppression of Ruby from CW - cheating will spread like a plague
all the more that everybody will now have a lot of ways to cheat. 
It will reduce to zero all the brilliant effort of the CW staff who made Codewars unique.

**Comment by danielpclark** [comment]
The definition of Kata is basically the exercise of disciplines (specific movements) to re-enforce the patterns you wish to master.  I believe some of you are over-reacting with the end-of-the-world for Ruby on CodeWars.

People like [Jim Weirich](https://www.youtube.com/watch?v=983zk0eqYLY) who demonstrate How a code Kata is to be done shows a place where he writes the code before the test... but then catches that he did that and deletes the all of the code he just did so he can write the test first and then write the code.  He's demonstrating proper kata technique.

Solving a solution by itself is not a Kata.  The practice of technique, with repetition, in bettering yourself is a proper Kata.

> I doubt very much that, from now on, somebody will write Ruby katas! 

I'm inclined to disagree with you.  Ruby is awesome in all regards and I've only seen those who have broken past my tests as an extra challenge for myself as a designer.  After all isn't that how software works in real life?  Designers and makers create and users do their own thing and then the designers refactor their own product.  That's just the life of software development.

> take a serious action

I'm welcome to whatever action he's willing to do and I won't be upset over it.

But g964; the defining of what a cheat is and it's acceptableness has not yet been described.  If you look at the thread discussing it you can see all points of view.  For example I think it's cheating when people look up the answer via Google to enter it in.  I'm even proud to say that I solved all of my solutions without doing any online searching for solutions but used my own knowledge to think and figure things out.  So my friend who has Googled his 1500 points... is that cheating more than my test passing via solving-myself the tests/challenges at hand?  Regardless it doesn't "feel" right to me that he does that, but that is perfect business development practice in the software world.

> cheating will spread like a plague

I highly doubt it.  As was discussed on the cheat-forum thread cheaters will be short lived and lose interest.  They will leave after they've obtained their mental image of what they want and they will not have improved themselves in the process; so in the end they won't feel the reward.

Again I want to emphasize what a Kata is and it's focus is on honing in on ordered disciplines to master an art.

@g964 in your point number 2 you come across as a bit hostile `serious action` with desperation `like a plague`.  Are you angry?

I appreciate bkaestner for his remarks in response to raulbc777 remarks about me crossing over the line.  I can understand both that the level of exposure this has is high and it would have been better to report the methodology and that there isn't a proper medium or policy to do so.  As the link raulbc777 has shared shows a user asking how my code works and I responded, with the spoiler flag, on every detail of Ruby that is used in the implementation.  He has found it very educational and appreciates my explaining it to him.

**Comment by bkaestner** [comment]
> @g964 in your point number 2 you come across as a bit hostile serious action with desperation like a plague. Are you angry?

I doubt that he's angry,  but instead hitting a language barrier, @danielpclark . Enabling `Annotations.for(g964)`:

> 2\. Protecting Ruby against cheaters is getting hard, and therefore, authors will lose interest in writing Ruby katas. If Jake doesn't act soon, cheating will spread like a spread like a plague, since the cheats are now known. One way to do that would be a penalty for obvious cheats.

> After all, cheating reduces the CW staff's effort to make Codewars unique.

English isn't g964's first language, and he's passionate in his original language, so although he seems "aggressive", he's more concerned.

**Comment by jhoffner** [comment]
Oh boy. So @danielpclark I went through your last 369 solutions, which seems to be where you cheating started, and marked all of those solutions as cheats. We have a penalty system in place to dock up to 40 points from those who cheat mor than 5 times. This means your new honor score came out to `-14122` :hankey:. I guess the negative points kind of act like a scarlet letter. Interestingly it only put you back down to a 4kyu, so I'll need to look into that more. 

I'm actually not sure what to do about leaving your score like that. Clearly its hopeless to try to recover from it. You cheated pretty openely and you were very frank in this thread about it which I applaud you for. I'm personally not in the "cheaters are evil" camp. Codewars is all about learning and exploration and you seemed to have learned some things by doing this. My biggest concern with cheating is just that for some users, the "gamification" is what keeps them driven to keep going, and we want everyone to stay driven so that they learn and get better. Clearly the real purpose for the site is about the intrinsic value of getting better at coding and not the extrinsic vaue of earning points. At the same time, even if it is all about the self-improvement, its kind of awesome to work hard and see yourself climb the ranks. There is something also to be said for its ability to act as an achievment on a CV. So... cheating does need to be handled when it happens. 

@g964 I wouldn't be too worried about rampant cheating. Almost everyone is on CW to get better and the dev community at large is a pretty honorable group. We have had loopholes in the system for a while, at one point there was a kata that was all about cheating, with solutions visible to everyone, and no one used them. In the end, cheaters are only cheating themselves. For those who do cheat, they eventually get caught, and in some cases end up with -14k honor.

In terms of writing kata that prevent this type of thing. IMO we are getting too far from the point if we need to write crazy cryptic test cases or all kinds of anti-hacking measures. If you want to add them to yours, cool, but the vision was always to make CW as close as possible to writing typical real-world code that you would write for your job. Even with random test cases, which is something a lot of authors use to protect their kata, no one really does that in practice in their actual software tests (though it does maybe have some real-world use).

**Comment by danielpclark** [comment]
Thanks @bkaestner and @jhoffner !  Having negative points won't deter me from enjoying the challenge and practice of Katas.  If you're okay with it, I'm okay with a negative value.

**Comment by g964** [comment]
I am sorry if I have seemed "angry" and I thank  @bkaestner for his "translation" of my concerns. 
I also thank @jhoffner for his good sense; I agree with what he wrote.
Finally I still think it would be interesting to have a little guide by @danielpclark on some unfamiliar 
(at least unfamiliar to me) faces of Ruby.
"All's well that ends well".

**Comment by jhoffner** [comment]
@bkaestner in terms of disclosure rules, I'll have to discuss with @nathandoctor the best official way to handle those, for now posting as an issue here or in our Gitter chat is fine. Gitter is a little bit less discoverable probably.

**Comment by raulbc777** [comment]
I have to thank  @jhoffner  for his action and concern about the situation we had here. I'm not happy with the result that @danielpclark had. We have to admit that the man is brilliant and he would had helped a lot to many users to detect the weaknesses of our tests, of course doings things completely different. I remember that @bkaestner prevented me, when I authored my first kata, that some users may use globals to catch some information about the solutions. I followed his instructions but to tell you the truth this is an area that I still need many things to learn. I will be creating more katas and I reaffirm my position: I won't accept a solution that uses a code designed with the intention of getting tests data and/ or passed the tests. To @danielpclark I want to say, he's not my enemy. He may be a great collaborator here and if  @jhoffner  considers to increase his negative score based on his attitude(the system reveals details) since now, I'll be happy with that, every body needs hope and a second chance, it's only a point of view.
